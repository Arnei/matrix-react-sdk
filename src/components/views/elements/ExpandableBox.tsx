/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classNames from "classnames";
import React, { useCallback, useState } from "react";

interface ExpandableBoxProps {
    body: React.ReactNode;
    className: string;// How many? 3?
    lines: number;// Number of lines to show before cut-off
}

/**
 * A container for text and inline elements. Initially limits its height to a number of lines.
 * If the text exceeds the number of lines, the full text can be shown via a toggle.
 */
const ExpandableBox: React.FC<ExpandableBoxProps> = ({ body, className, lines = 2 }) => {
    const [lineHeight, setLineHeight] = useState(16);       // Default replaced by computedStyle value
    const [scrollHeight, setScrollHeight] = useState(32);   // Default replaced in ref callback
    const [isExpanded, setIsExpanded] = useState(false);
    const [isOverflow, setIsOverflow] = useState(false);

    const ToggleButton = ({ isExpanded, onClick }) => {
        return (
            <button className="mx_RoomSummaryCard_infoTopic_toggle" onClick={onClick}>
                { isExpanded ? "Less" : "More" }
            </button>
        );
    };

    const handleToggle = () => {
        setIsExpanded((prev) => !prev);
    };

    const checkScrollHeight = useCallback((node: HTMLDivElement) => {
        if (node != null) {
            const computedLineHeight = parseInt(getComputedStyle(node).getPropertyValue("line-height"));
            const maxHeight = computedLineHeight * lines;
            setLineHeight(computedLineHeight);
            setScrollHeight(node.scrollHeight);

            if (node.scrollHeight > maxHeight) {
                setIsOverflow(true);
            }
        }
    // Run when body changes to check if we still overflow
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [body]);

    // Use styles to expand or shrink the component based on the toggle
    const styleNotExpanded = {
        lineHeight: `${lineHeight}px`,
        WebkitLineClamp: `${lines}`,
    };

    const styleExpanded = {
        height: `${scrollHeight}px`,
        lineHeight: `${lineHeight}px`,
        WebkitLineClamp: 'unset',
    };

    return (
        <div className="mx_RoomSummaryCard_infoTopic">
            <div ref={checkScrollHeight}
                className={classNames("mx_RoomSummaryCard_infoTopic_text", className)}
                style={isExpanded ? styleExpanded : styleNotExpanded}>
                { body }
            </div>
            { isOverflow && <ToggleButton isExpanded={isExpanded} onClick={handleToggle} /> }
        </div>
    );
};

export default ExpandableBox;
