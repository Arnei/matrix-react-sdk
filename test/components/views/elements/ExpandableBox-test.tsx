/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
// eslint-disable-next-line deprecate/import
import { mount } from 'enzyme';

import ExpandableBox from "../../../../src/components/views/elements/ExpandableBox";

describe('<ExpandableBox />', () => {
    const defaultProps = {
        body: "Here is a bit of text.",
        className: 'mx_Icon_16',
        lines: 3,
    };

    const getComponent = (props = {}) =>
        mount(<ExpandableBox {...defaultProps} {...props} />);

    it('renders at all', () => {
        const component = getComponent();
        expect(component).toMatchSnapshot();
    });

    it('renders no button if not overflowing', () => {
        const component = getComponent();
        expect(component.find('.mx_RoomSummaryCard_infoTopic_toggle').exists()).toBeFalsy();
    });

    it('renders button if overflowing', () => {
        const component = getComponent({ lines: -1 });
        expect(component).toMatchSnapshot();
        expect(component.find('.mx_RoomSummaryCard_infoTopic_toggle').exists()).toBeTruthy();
    });

    it('toggle button toggles style', () => {
        const component = getComponent({ lines: -1 });

        component.find('.mx_RoomSummaryCard_infoTopic_toggle').simulate('click');

        let webkitLineClampfff = (component.find('.mx_RoomSummaryCard_infoTopic_text').getDOMNode<HTMLElement>()
            .style as any).WebkitLineClamp;
        expect(webkitLineClampfff).toEqual("unset");

        component.find('.mx_RoomSummaryCard_infoTopic_toggle').simulate('click');

        webkitLineClampfff = (component.find('.mx_RoomSummaryCard_infoTopic_text').getDOMNode<HTMLElement>()
            .style as any).WebkitLineClamp;
        expect(webkitLineClampfff).toEqual("-1");
    });

    // TODO: Somehow restrict the width of the component to get the ToggleButton to render
    // and then test it.
});
