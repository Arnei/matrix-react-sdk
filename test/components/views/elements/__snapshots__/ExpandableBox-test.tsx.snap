// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpandableBox /> renders at all 1`] = `
<ExpandableBox
  body="Here is a bit of text."
  className="mx_Icon_16"
  lines={3}
>
  <div
    className="mx_RoomSummaryCard_infoTopic"
  >
    <div
      className="mx_RoomSummaryCard_infoTopic_text mx_Icon_16"
      style={
        Object {
          "WebkitLineClamp": "3",
          "lineHeight": "16px",
        }
      }
    >
      Here is a bit of text.
    </div>
  </div>
</ExpandableBox>
`;

exports[`<ExpandableBox /> renders button if overflowing 1`] = `
<ExpandableBox
  body="Here is a bit of text."
  className="mx_Icon_16"
  lines={-1}
>
  <div
    className="mx_RoomSummaryCard_infoTopic"
  >
    <div
      className="mx_RoomSummaryCard_infoTopic_text mx_Icon_16"
      style={
        Object {
          "WebkitLineClamp": "-1",
          "lineHeight": "16px",
        }
      }
    >
      Here is a bit of text.
    </div>
    <ToggleButton
      isExpanded={false}
      onClick={[Function]}
    >
      <button
        className="mx_RoomSummaryCard_infoTopic_toggle"
        onClick={[Function]}
      >
        More
      </button>
    </ToggleButton>
  </div>
</ExpandableBox>
`;
